tests/strands/tools/test_registry.py

```diff
@@ -260,3 +260,148 @@ def test_register_strands_tools_module_non_callable_function():
         " Tool tool_with_spec_but_non_callable_function function is not callable",
     ):
         tool_registry.process_tools(["tests.fixtures.tool_with_spec_but_non_callable_function"])
+
+
+def test_tool_registry_cleanup_with_mcp_client():
+    """Test that ToolRegistry cleanup properly handles MCP clients without orphaning threads."""
+    from unittest.mock import AsyncMock, MagicMock
+
+    from strands.tools.mcp import MCPClient
+
+    # Create a mock MCP client that simulates a real tool provider
+    mock_transport = MagicMock()
+    mock_client = MCPClient(mock_transport)
+
+    # Mock the client to avoid actual network operations
+    mock_client.load_tools = AsyncMock(return_value=[])
+
+    registry = ToolRegistry()
+
+    # Use process_tools to properly register the client
+    registry.process_tools([mock_client])
+
+    # Verify the client was registered as a consumer
+    assert registry._registry_id in mock_client._consumers
+
+    # Test cleanup calls remove_consumer
+    registry.cleanup()
+
+    # Verify cleanup was attempted
+    assert registry._registry_id not in mock_client._consumers
+
+
+def test_tool_registry_cleanup_exception_handling():
+    """Test that ToolRegistry cleanup attempts all providers even if some fail."""
+    from unittest.mock import MagicMock
+
+    # Create mock providers - one that fails, one that succeeds
+    failing_provider = MagicMock()
+    failing_provider.remove_consumer.side_effect = Exception("Cleanup failed")
+
+    working_provider = MagicMock()
+
+    registry = ToolRegistry()
+    registry._tool_providers = [failing_provider, working_provider]
+
+    # Cleanup should attempt both providers and raise the first exception
+    with pytest.raises(Exception, match="Cleanup failed"):
+        registry.cleanup()
+
+    # Verify both providers were attempted
+    failing_provider.remove_consumer.assert_called_once()
+    working_provider.remove_consumer.assert_called_once()
+
+
+def test_tool_registry_cleanup_idempotent():
+    """Test that ToolRegistry cleanup is idempotent."""
+    from unittest.mock import AsyncMock, MagicMock
+
+    from strands.experimental.tools import ToolProvider
```

Comment:
When we make these non-experimental, let's make these top-level imports throughout the file to avoid the repetition